#! /usr/bin/env Rscript

library(argparse)
options(stringsAsFactors=F)
parser=ArgumentParser()
parser$add_argument("-i", "--infile", help="Input QIIME key")
parser$add_argument("-o", "--outfile", help="Output QIIME key")
parser$add_argument("-a", "--alphadiv", help="File of alpha diversity metrics to add (specifically, observed_otus, chao1, shannon, and faith's PD after rarefaction to 5000 samples)")
parser$add_argument("-b", "--barcodes", help="File of Illumina barcodes for each sample")
parser$add_argument("-c", "--opencounts", help="Report of observation counts per sample from open-reference BIOM file (generated by biom summarize-table command)")
parser$add_argument("-q", "--closedcounts", help="Report of observation counts per sample from closed-reference BIOM file (generated by biom summarize-table command)")
parser$add_argument("-m", "--median-lengths", help="File of median read lengths for all samples")
parser$add_argument("-s", "--seqcounts", help="Total sequences after joining paired-end reads")
args=parser$parse_args()
# setwd('/home/jgwall/Projects/282MaizeLeaf16s_cleaned/2_QiimeOtus/')
#args=parser$parse_args(c("-i","2f_otu_table.sample_filtered.no_mitochondria_chloroplast.key.tsv", "-o", "99_tmp", '-b', '2h_barcode_key.txt', '-a', '2h_alpha_diversity_key.txt', 
#   '-c', '2f_otu_table.sample_filtered.no_mitochondria_chloroplast.samples.txt', '-s', '../1_ParseRawSequence/1b_read_counts.txt', '-m', '../1_ParseRawSequence/1c_median_lengths.txt', 
#   '-q', '2h_closed_reference_otu_table.samples.txt'))

# Load data
cat("Adding Earth Microbiome Project metadata to keyfile\n")
key=read.delim(args$infile, check.names=F)
barcodes=read.delim(args$barcodes, header=F, col.names=c("sample","barcode"))
alpha = read.delim(args$alphadiv, row.names=1)
   alpha$sample=rownames(alpha)
opencounts=read.table(args$opencounts, skip=15, header=F, col.names=c('sample', 'count'))
   opencounts$sample=sub(opencounts$sample, pattern=":", repl="", fixed=T)
closedcounts=read.table(args$closedcounts, skip=15, header=F, col.names=c('sample', 'count'))
   closedcounts$sample=sub(closedcounts$sample, pattern=":", repl="", fixed=T)
seqcounts=read.delim(args$seqcounts) 
   seqcounts$sample = gsub(seqcounts$sample, pattern="_", repl=".")
medians=read.delim(args$median_lengths) 
   medians$sample = gsub(medians$sample, pattern="_", repl=".")

# Helper function to make sure I'm adding things in the right order; assumes a column named 'sample'
add_with_verification=function(data, toadd, column){
  keepers = !is.na(toadd$sample)
  if(!identical(data$'#SampleID'[keepers], toadd$sample[keepers])){
	stop("Columns do not match when trying to add ",column)
  }
  return(toadd[[column]])
}

# Metadata - Sample
barcodematch = match(key$'#SampleID', barcodes$sample)
cat("\tMatched barcodes to", sum(is.finite(barcodematch)), "out of", nrow(key),"samples\n")
key$BarcodeSequence = add_with_verification(key, barcodes[barcodematch,], "barcode")
key$LinkerPrimerSequence = ""

# Metadata - Study
key$title="Quantitative Genetics of the Maize Phyllosphere"
key$principal_investigator="Jason G. Wallace"

# Metadata - Preparation
key$target_gene = "16S rRNA"
key$target_subfragment = "V5/V6"
key$pcr_primers = "FWD: ACCMGGATTAGATACCCKG; REV: AGGGTTGCGCTCGTTG"
key$illumina_technology = "HiSeq2500"
key$run_center = "BRC-Cornell"
key$run_date = "28 July 2015"

# Metadata - Sequences
# # Median read length in each library
medmatch = match(key$'#SampleID', medians$sample)
cat("\tMatched median read lengths to", sum(is.finite(medmatch)), "out of", nrow(key),"samples\n")
key$read_length_bp = add_with_verification(key, medians[medmatch,], "median_length")
# # Number of reads in each library before OTU calling
seqmatch = match(key$'#SampleID', seqcounts$sample)
cat("\tMatched split-library sequence counts to", sum(is.finite(seqmatch)), "out of", nrow(key),"samples\n")
key$sequences_split_libraries = add_with_verification(key, seqcounts[seqmatch,], "reads")
# # Closed-reference OTU counts
closedmatch = match(key$'#SampleID', closedcounts$sample)
cat("\tMatched closed-reference sequence counts to", sum(is.finite(closedmatch)), "out of", nrow(key),"samples\n")
key$observations_closed_ref_greengenes = add_with_verification(key, closedcounts[closedmatch,], "count")
# # Open-reference OTU counts
openmatch = match(key$'#SampleID', opencounts$sample)
cat("\tMatched open-reference sequence counts to", sum(is.finite(openmatch)), "out of", nrow(key),"samples\n")
key$observations_open_ref_greengenes = add_with_verification(key, opencounts[openmatch,], "count")


# Metadata - Taxonomy
key$sample_taxid="NA"
key$sample_scientific_name="NA"
key$host_taxid=381124
key$host_common_name_provided="Maize"
key$host_common_name="maize"
key$host_scientific_name="Zea mays subsp. mays"
key$host_superkingdom="sk__Eukaryota"
key$host_kingdom="k__Viridiplantae"
key$host_phylum="p__Streptophyta"
key$host_class = "c__Liliopsida"
key$host_order ="o__Poales"
key$host_family ="f__Poaceae"
key$host_genus = "g__Zea"
key$host_species="s__Zea_mays"


# Metadata - Geography
key$collection_timestamp[key$tissue_date=="LMAD_8"] = "2014-08-08 11:00:00"
key$collection_timestamp[key$tissue_date=="LMAN_8"] = "2014-08-08 23:00:00"
key$collection_timestamp[key$tissue_date=="LMAD_26"] = "2014-08-26 11:00:00"
key$collection_timestamp[key$tissue_date=="LMAN_26"] = "2014-08-26 23:00:00"
key$country="GAZ:United States of America"
key$latitude_deg=42.724554
key$longitude_deg=-76.653081
key$depth_m=0
key$altitude_m=1.5
key$elevation_m=260


# Metadata - ontology
key$env_biome="cropland biome"
key$env_feature="agricultural feature"
key$env_material="leaf"
key$envo_biome_0="biome"
key$envo_biome_1="terrestrial biome"
key$envo_biome_2="anthropogenic terrestrial biome"
key$envo_biome_3="cropland biome"
key$envo_biome_4=""
key$envo_biome_5=""
key$empo_0=""
key$empo_1="Host-associated"
key$empo_2="Plant"
key$empo_3="Plant surface"


# Metadata - Alpha-diversity
alphamatch = match(key$'#SampleID', alpha$sample)
cat("\tMatched alpha diversity to", sum(is.finite(alphamatch)), "out of", nrow(key),"samples\n")
key$adiv_observed_otus = add_with_verification(key, alpha[alphamatch,], "observed_otus")
key$adiv_chao1 = add_with_verification(key, alpha[alphamatch,], "chao1")
key$adiv_shannon = add_with_verification(key, alpha[alphamatch,], "shannon")
key$adiv_faith_pd = add_with_verification(key, alpha[alphamatch,], "PD_whole_tree")

# The Subsets and Physiochemical metadata sets don't make sense for this data

# Final cleanups
key$date=NULL	# Duplicated by timestamp
key$Description=paste("MaizeInbred.", key$Description, sep="")
key$subpop[key$subpop=="nss"] = "non_stiff_stalk"
key$subpop[key$subpop=="ss"] = "stiff_stalk"
key$subpop[key$subpop=="ts"] = "tropical_subtropical"
key$subpop[key$subpop=="sweet"] = "sweetcorn"

# write out reorganized version
write.table(key, file=args$outfile, sep='\t', quote=F, row.names=F, col.names=T)